File: CSharp/PracticeSet2/Backtracking/SubSets.cs
/*
Example:

Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
*/
public class Solution {
    public IList<IList<int>> Subsets(int[] nums) {
 		IList<IList<int>> result = new List<IList<int>>();
 		if(nums.Length==0)
 		{
 			return result;
 		}
 		IList<int> innerList = new List<int>();
 		Helper(nums,0,innerList,result);
 		return result;
    }
    private void Helper(int[] nums, int index, IList<int> innerList, IList<IList<int>> result)
    {
    	if(index<nums.Length)
    	{
    		Helper(nums,index+1,innerList,result);
    		IList<int>tempList = new List<int>();
    		for(int i=0;i<innerList.Count;i++)
    		{
    			tempList.Add(innerList[i]);
    		}
    		tempList.Add(nums[index]);
    		Helper(nums,index+1,tempList,result);
    	}
    	else
    	{
    		result.Add(innerList);
    	}
    }
}

File: Java/Moderate/Backtracking/Subsets.java
/*
Input: nums = [1,2,3]
Output:
[
  [3],
  [1],
  [2],
  [1,2,3],
  [1,3],
  [2,3],
  [1,2],
  []
]
*/
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if(nums.length==0)
        {
        	return result;
        }
        
        for(int i=0;i<nums.length;i++)
        {
          List<List<Integer>> temp = new ArrayList<List<Integer>>();
          for(int j=0;j<result.size();j++)
          {
            temp.add(new ArrayList<Integer>(result.get(j)));
          }

          for(int j=0;j<temp.size();j++)
          {
            temp.get(j).add(nums[i]);
          }

          List<Integer> singleList = new ArrayList<Integer>();
          singleList.add(nums[i]);
          result.add(singleList);
          result.addAll(temp);
        }
        result.add(new ArrayList<Integer>());
        return result;
    }
}